### ISSUE TITLE
Inconsistent results from cabal build and cabal install

### ISSUE THREAD

**Issue by recursion-ninja** [haskell/cabal#7210]
**Describe the bug**
When `cabal build` successfully constructs all build targets but `cabal install` fails to construct one or more build targets.

**To Reproduce**

```
$ git clone https://github.com/amnh/PCG
$ cd PCG
$ git checkout f3e5a1b91fbfe37f3e17c322cab7fe696541a110
$ cabal v2-build   # this succeeds (expected)
$ cabal v2-install # this fails    (unexpected)
```

**Expected behavior**
I expect that both `cabal build` and `cabal install` will successfully construct all build targets

**System information**
 - Operating system
    ```
    $ lsb_release -a
    No LSB modules are available.
    Distributor ID: Ubuntu
    Description:    Ubuntu 20.04.1 LTS
    Release:        20.04
    Codename:       focal
   ```
 - Cabal
   ```
   $ cabal --version
   cabal-install version 3.4.0.0`
   compiled using version 3.4.0.0 of the Cabal library
   ```
 - GHC 
   ```
   $ ghc --version
   The Glorious Glasgow Haskell Compilation System, version 8.10.3
    ```

**Additional context**

It appears that there is a discrepancy with the way the FFI is being handled between `cabal build` and `cabal install`. Specifically, the `tcm-memo` build target succeeds with `cabal build` and fails with `cabal install`. It com plains about missing C and C++ header files. if these header files are are added to the `.cabal` file under `c-sources` and `cxx-sources` the `cabal install` build will progress slightly before failing with the message `ghc: no input files`.  It is also worth noting that if the C and C++ header files are added to the `.cabal` file, `cabal build` will now fail to build `tcm-memo` for different reasons than `cabal install`, leading me to think that the C and C++ hear files should not be explicitly added. I suspect there is an issue somewhere with how `hsc2hs` is being implicitly invoked, but the specifics are beyond my scrutiny. 

I would greatly appreciate help in understanding this behavior and resolving any discrepancies between `cabal build` and `cabal install`.

**Comment by fgaz** [comment]
`cabal install` does a sdist before building the package. Does the error still occur if you do `cabal sdist`, then unpack the .tar.gz in a separate directory and `cabal build` that?

**Comment by recursion-ninja** [comment]
@fgaz Thank you for advising me on how the `sdist` command is interrelated to the `install` command. After inspecting the unpacked results of `cabal sdist` I found that the C and C++ header files were missing, they were not bundled when `sdist` was run. I added the C and C++ header files to the `extra-source-files` stanza of the `.cabal` file. This caused the header files to be correctly bundled with `cabal sdist` and consequently fixed the build errors with `cabal install`.

*Note: Do not add the C and C++ header files to the `c-sources` and `cxx-sources` stanzas recpectively as this causes build errors with `cabal install` and `cabal build`.*
